lib/Runtime/Library/JavascriptObject.cpp

```diff
@@ -357,7 +357,127 @@ namespace Js
+        if (type == TypeIds_Undefined)
+        {
+            return library->CreateStringFromCppLiteral(_u("[object Undefined]"));
+        }
+        // 2. If the this value is null, return "[object Null]".
+        if (type == TypeIds_Null)
+        {
+            return library->CreateStringFromCppLiteral(_u("[object Null]"));
+        }
+
+        // 3. Let O be ToObject(this value).
+        RecyclableObject *thisArgAsObject = RecyclableObject::FromVar(JavascriptOperators::ToObject(thisArg, scriptContext));
+
+        // 4. Let isArray be ? IsArray(O).
+        // There is an implicit check for a null proxy handler in IsArray, so use the operator.
+        BOOL isArray = JavascriptOperators::IsArray(thisArgAsObject);
+
+        Var tag = nullptr;
+        // 15. Let tag be ? Get(O, @@toStringTag).
+        try
+        {
+            tag = JavascriptOperators::GetProperty(thisArgAsObject, PropertyIds::_symbolToStringTag, scriptContext); // Let tag be the result of Get(O, @@toStringTag).
+        }
+        catch (JavascriptExceptionObject*)
+        {
+            tag = thisArgAsObject;
+        }
+
+        // 17. Return the String that is the result of concatenating "[object ", tag, and "]". 
+        if (tag != nullptr && JavascriptString::Is(tag))
+        {
+            JavascriptString *tagStr = JavascriptString::FromVar(tag);
+
+            CompoundString::Builder<32> stringBuilder(scriptContext);
+
+            stringBuilder.AppendChars(_u("[object "));
+
+            stringBuilder.AppendChars(tagStr);
+            stringBuilder.AppendChars(_u(']'));
+
+            return stringBuilder.ToString();
+        }
+
+        // If we don't have a tag or it's not a string, use the 'built in tag'.
+        if (isArray)
+        {
+            // 5. If isArray is true, let builtinTag be "Array".
+            return library->CreateStringFromCppLiteral(_u("[object Array]"));
+        }
+
+        JavascriptString* builtInTag = nullptr;
+        switch (type)
+        {
+            // 6. Else if O is an exotic String object, let builtinTag be "String".
+        case TypeIds_String:
+        case TypeIds_StringObject:
+            builtInTag = library->CreateStringFromCppLiteral(_u("[object String]"));
+            break;
+
+            // 7. Else if O has an[[ParameterMap]] internal slot, let builtinTag be "Arguments".
+        case TypeIds_Arguments:
+            builtInTag = library->CreateStringFromCppLiteral(_u("[object Arguments]"));
+            break;
+
+            // 8. Else if O has a [[Call]] internal method, let builtinTag be "Function".
+        case TypeIds_Function:
+            builtInTag = library->CreateStringFromCppLiteral(_u("[object Function]"));
+            break;
+
+            // 9. Else if O has an [[ErrorData]] internal slot, let builtinTag be "Error".
+        case TypeIds_Error:
+            builtInTag = library->CreateStringFromCppLiteral(_u("[object Error]"));
```

Comment:
It looks like that isn't used anywhere. Maybe I can remove it from JavascriptLibrary.

GetObjectDisplayString is similar. After this change, shouldn't it be argued that anything using this should have used EntryToString? We have a use in ToLocaleString on a VariantDate.